README.md
[![BitBox02 Logo](docs/assets/bitbox02.png)](https://github.com/digitalbitbox/bitbox02-api-js/blob/master/docs/methods.md)

# BitBox02 JavaScript library

The JavaScript library is compiled from
[bitbox02-api-go](https://github.com/digitalbitbox/bitbox02-api-go) using GopherJS.

Given that it contains the full client implementation compiled from Go, the library is quite large (~3MB).
We recommend lazy-loading it only when necessary.

## BitBox02

The BitBox02 is a hardware wallet made by [Shift Crypto](https://shiftcrypto.ch) that simplifies secure handling of crypto coins through storing private keys and signing transactions.

## Integration

To enable communication from the browser to the BitBox02, either:
  - WebHID needs to be supported by the browser, e.g. Chrome
  - or the [BitBoxBridge](https://github.com/digitalbitbox/bitbox-bridge) needs to be installed and running

When integrating the BitBox02 into your application, your domain needs to be whitelisted in the BitBoxBridge.
To do so, please submit a Pull Request or an Issue in the [bitbox-bridge](https://github.com/digitalbitbox/bitbox-bridge) repository.
Localhost is already whitelisted, so you can develop locally.

The BitBox02 Javascript library is available as NPM package [bitbox02-api](https://www.npmjs.com/package/bitbox02-api).

### Install

```sh
$ npm install bitbox02-api
```

### Import

```javascript
import { BitBox02API, getDevicePath} from 'bitbox02-api';
```

### Initialize

To get the device path, the BitBoxBridge needs to be running.
If WebHID (`navigator.hid`) is available, `getDevicePath()` returns `"WEBHID"`, which instructs `BitBox02API` to prefer WebHID over the BitBoxBridge.

```javascript
const devicePath = await getDevicePath();
const BitBox02 = new BitBox02API(devicePath);
```

### Connect to device

The `BitBox02API.connect()` method takes 5 arguments:

```javascript
/**
 * @param showPairingCb Callback that is used to show pairing code. Must not block.
 * @param userVerify Promise that should resolve once the user wants to continue.
 * @param handleAttastionCb Callback that should handle the bool attestation result. Must not block.
 * @param onCloseCb Callback that's called when the websocket connection is closed.
 * @param setStatusCb Callback that lets the API set the status received from the device.
 * @return Promise that will resolve once the pairing is complete.
 */
connect (showPairingCb, userVerify, handleAttestationCb, onCloseCb, setStatusCb)
```

### Check BitBox02 edition

The BitBox02 is available in two editions: "Multi" and "Bitcoin-only".
To check what edition is used, e.g. to make sure that Ethereum functionality is supported, use the following function.

```javascript
import { constants } from 'bitbox02-api';

BitBox02.firmware().Product() === constants.Product.BitBox02Multi
BitBox02.firmware().Product() === constants.Product.BitBox02BTCOnly
```

### Methods

All available methods are documented in [`docs/methods.md`](docs/methods.md).

## Sample integration

This is a sample BitBox02Wallet class integration for connecting to the BitBox02 device using the BitBoxBridge and this JS API.
See [`sandbox/src/index.js`](sandbox/src/index.js) for a fully functional sandbox implementation.

```javascript
import {
  constants,
  BitBox02API,
  getDevicePath,
  sanitizeEthTransactionData
} from 'bitbox02-api';

class BitBox02 {
  constructor(logout) {  // You can provide the `logout` callback of your application in the constructor
    this.logout = logout;
    this.status = undefined;
    this.pairingConfirmed = false;
  }

  async init(keypath) {
    try {
      const devicePath = await getDevicePath();
      this.api = new BitBox02API(devicePath);

      await this.api.connect(

        /** @param showPairingCb
         *  Store the pairing code on the class instance. Show this to the user to compare with code
         *  on the device when `this.status === 'unpaired'`
         */
        pairingCode => {
          this.pairingCode = pairingCode;
        },

        /** @param userVerify
         *  Store the Promise's `resolve` on the class instance to call when the user clicks the corresponding button
         *  in your application after confirming the pairing on device
         */
        async () => {
          return new Promise(resolve => {
            this.pairingConfirmed = true;
            this.pairingConfirmationResolve = resolve;
          });
        },

        /** @param handleAttestationCb
        *  Store the attestation result on the class instance. If attestation fails, the user might have a fake device.
        *  Handle this condition below.
        */
        attestationResult => {
          this.attestation = attestationResult;
        },

        /** @param onCloseCb
        *  Log the user out of your application when device is unplugged/the websocket closes.
        *  Here we use the `logout` function provided in the constructor as the callback.
        */
        () => {
          this.logout();
        },

        /** @param setStatusCb
        *  Store the status on the class instance to take appropriate actions based on status.
        *  All possible status can be found here: https://github.com/digitalbitbox/bitbox02-api-go/blob/master/api/firmware/status.go
        */
        status => {
          this.status = status;
        }
      );
    } catch(e) {
      alert(e);
      this.logout();
      return;
    }

    switch (this.api.firmware().Product()) {
        case constants.Product.BitBox02Multi:
            console.log("This is a BitBox02 Multi");
            break;
        case constants.Product.BitBox02BTCOnly:
            console.log("This is a BitBox02 BTC-only");
            break;
    }

    // Handle attestattion failure
    if (!this.attestation) {
      alert('Attestation failed');
    }

  }
}

const device = new BitBox02(yourLogoutFunction)
await device.init()

// Now you can call any of the supported API methods documented above e.g.:
const ethPub = await device.api.ethGetRootPubKey("m/44'/60'/0'/0");

```

## Local development

To compile the library using GopherJS, run the following commands:

```sh
$ make dockerinit
$ make dockercompile
```

To simply run the demo sandbox implementation, run the following command and visit <http://localhost:8000>.

```sh
$ make builddemo
$ make servedemo
```


CHANGELOG.md
# 0.15.1
- `ethSignTypedMessage()` now accepts hex strings (e.g. `"0x01"`) for the `uint` types
- re-enable minification of the bitbox02-api-go.js

# 0.15.0
- Add `ethSignTypedMessage()` (EIP-712 support)

# 0.14.0

- Allow large ETH chain IDs
- bitbox02-api-go: updated to 67b3f707055c6b49ca9491b6c0d9003ad0f0cc10

# 0.13.0
- Cardano: allow a TTL value of "0" (before "0" would be treated the same as undefined/missing).
- Cardano: add support for tokens

# 0.12.0
- Fix WebHID vendor/product filters, fixing WebHID access in Chrome 97+
- Breaking change: rename `hash` to `payload` in Bitcoin outputs in `btcSignSimple()` and `btcSignMultisig()`

# 0.11.0
- Converted bitbox02-api-js to a CommonJS module for better integration with webpack
- Converted the demo to a webpack project

# 0.10.0
- Added Cardano support
- bitbox02-api-go: updated to 93df9155c1466ae51fbe1ce88a963a9ee709070f
- add function to retrieve the firmware version

# 0.9.0

- Updated Go to version 1.16
- Drop requirement for running the BitBoxBridge if WebHID is available in the Browser (Chrome only for now)
- Publish transpiled ES5 code instead of source ES6 code for better compatibility

# 0.8.1
- Added support for display=false argument for ethDisplayAddress and btcDisplayAddressSimple
- Changed demo to show address first and then ask for user confirmation on the BitBox
- Fixed ethDisplayAddress does not return address
- Updated documentation for btcSignMessage

# 0.8.0
- bitbox02-api-go: updated to 3cdcbc6489771e54fdfa18a205416ab1eee291cb
- Added btcSignMessage()

# 0.7.0:
- bitbox02-api-go: updated to f23403837ec161298c08a2e5286bc62f043c3c48

# 0.6.0
- bitbox02-api-go: updated to 753a8f6ca53ee056dcbf6e629c5335096ae116fb (v9.0.0 compatibility)

# 0.5.0
- bitbox02-api-go: updated to f9fe88be0a9b8866533a20d01d264fe42a085fd9 (v8.0.0 compatibility)

# 0.4.1
- Updated BitBoxSandbox demo
   - Add methods documentation in docs/methods.md
   - Sandbox UI improvements

# 0.4.0
- Changed api.firmware.* to constants.* (all constants)
- Changed api.firmware.isErrorAbort to isErrorAbort (direct export)
- api no longer exported (treated as internal).
- add btcXPub, btcMaybeRegisterScriptConfig, btcDisplayAddressMultisig, btcSignMultisig api calls.
- bitbox02-api-go: updated to 49d314b148cccb193e26b5c154044ed7eb819384 (v7.0.0 compatibility)
- Add Jest framework and Travis CI for testing and add unit tests for util functions
- Improve error handling with regards to CORS headers when connecting to the device. We can now correctly identify when the origin is not whitelisted based on the response from the bridge

# 0.3.0
- Update from Go API: https://github.com/digitalbitbox/bitbox02-api-go/commit/7817ceb1fccc7276f17a6a468221258cbd4d5bac
    - Includes support for firmware 6.1.0 and makes it the lowest supported version
- Update sandbox demo to use a class and abstracted ethereum api methods
- Return arrays instead of Buffers in ethereum signature results and let the caller handle conversion as they require
- Update README and docstrings
- Access `firmwareAPI` more safely by checking if connection is valid first
- Make `AsyncETHSign` in the go wrapper take []byte for all arguments which allows us to completely remove `sanitizeEthTransactionData`
    - Update README and `demo.js` accordingly to reflect these changes
    - TODO: Mirror these changes in ETHSign in the go client and the BitBoxApp to handle Eth TXs the same everywhere

# 0.2.0

- Add CHANGELOG
- Update README
- Add AsyncBTCAddressSimple, AsyncBTCSignSimple
- Add `getCoinFromChainId()`, `getKeypathFromString()`, and `getCoinFromKeypath()` eth utils to handle mainnet and testnet keypaths and coins correctly. Currently only two keypaths are supported:
    - `m/44'/60'/0'/0` for mainnet and
    - `m/44'/1'/0'/0`  for Rinkeby and Ropsten testnets


# 0.1.0

- Publish first API version


demo/demo.js
import { isErrorAbort, constants, getDevicePath, BitBox02API, getKeypathFromString, HARDENED } from './bitbox02-api.js'

// data: Uint8Array.
function toBase64(data) {
    return window.btoa(String.fromCharCode.apply(null, data));
}

// data: Uint8Array.
function toHex(data) {
    return [...data].map(b => b.toString(16).padStart(2, "0")).join("");
}

function fromHex(hexString) {
    return new Uint8Array(hexString.match(/.{1,2}/g).map(byte => parseInt(byte, 16)));
}

function reset() {
    document.getElementById("demo").disabled = false;
    document.getElementById("pairing").style.display = "none";
    document.getElementById("pairingOK").disabled = true;
    document.getElementById("initialized").style.display = "none";
    document.getElementById("intro").style.display = "flex";
}

class BitBox02 {
    constructor(logout) {
        this.logout = logout;
    }

    async init() {
        const pairingOKButton = document.getElementById("pairingOK");
        const initializedDiv = document.getElementById("initialized");
        document.getElementById("demo").disabled = true;
        try {
            const devicePath = await getDevicePath();
            this.api = new BitBox02API(devicePath);

            await this.api.connect(
                pairingCode => {
                    document.getElementById("intro").style.display = "none";
                    document.getElementById("pairing").style.display = "flex";
                    document.getElementById("pairingCode").innerHTML = pairingCode.replace("\n", "<br/>");
                },
                () => new Promise(resolve => {
                    pairingOKButton.disabled = false;
                    pairingOKButton.addEventListener("click", resolve);
                }),
                attestationResult => {
                    console.log("Attestation check:", attestationResult);
                },
                this.logout,
                status => {
                    console.log(status);
                }
            )
        } catch (e) {
            alert(e);
            reset();
            return;
        }
        switch (this.api.firmware().Product()) {
            case constants.Product.BitBox02Multi:
                console.log("This is a BitBox02 Multi");
                let elements = document.getElementsByClassName("bb02-multi");
                for (let i = 0; i < elements.length; i++) {
                    elements[i].disabled = false;
                }
                break;
            case constants.Product.BitBox02BTCOnly:
                console.log("This is a BitBox02 BTC-only");
                break;
        }

        document.getElementById("pairing").style.display = "none";
        document.getElementById("intro").style.display = "none";
        initializedDiv.style.display = "flex";
    }
}

let device;
const runDemo = async () => {
    device  = new BitBox02(reset)
    await device.init();
}

// ---- Demo buttons

// Start the demo
document.getElementById("demo").addEventListener("click", runDemo);

// Close the connection (logout).
document.getElementById("close").addEventListener("click", () => {
    device.api.close();
});


// Get ethereum xpub for given keypath
ethPub.addEventListener("click", async () => {
    const ethPub = await device.api.ethGetRootPubKey("m/44'/60'/0'/0");
    alert(ethPub);
});

// Get ethereum address for given keypath
// Only displays address on device, does not return. For verification, derive address from xpub
const ethAddrBtn = document.querySelector("#ethAddr");
ethAddrBtn.addEventListener("click", async () => {
    const address = await device.api.ethDisplayAddress("m/44'/60'/0'/0/0", false);
    ethAddrBtn.textContent = "Confirm the following address on the BitBox:\n" + address;
    ethAddrBtn.className = "btn btn-warning";
    try {
        const addressConfirmed = await device.api.ethDisplayAddress("m/44'/60'/0'/0/0");
        ethAddrBtn.textContent = "Success!:\n" + addressConfirmed;
        ethAddrBtn.className = "btn btn-success";
    } catch (e) {
        ethAddrBtn.textContent = JSON.stringify(e);
        ethAddrBtn.className = "btn btn-danger";
    }
});

// Sign ethereum transaction
ethSign.addEventListener("click", async () => {
    const signingData = {
        keypath: "m/44'/60'/0'/0/0", // mainnet tx needs a mainnet keypath
        chainId: 1,                  // mainnet tx
        tx: {
            nonce: new Uint8Array([0]),
            gasPrice: new Uint8Array([76, 153, 237, 154, 0]),
            gasLimit: new Uint8Array([82, 8]),
            to: new Uint8Array([4, 242, 100, 207, 52, 68, 3, 19, 180, 160, 25, 42, 53, 40, 20, 251, 233, 39, 184, 133]),
            value: new Uint8Array([6, 240, 91, 89, 211, 178, 0, 0]),
            data: new Uint8Array([])
        }
    }
    try {
        const sig = await device.api.ethSignTransaction(signingData);
        console.log(sig);
    } catch(e) {
        alert(e);
    }
});

// Sign "hello world" ethereum message
ethSignMsg.addEventListener("click", async () => {
    try {
        const sig = await device.api.ethSignMessage({
            keypath: "m/44'/60'/0'/0/0",
            // "hello world"
            message: new Uint8Array([104, 101, 108, 108, 111, 32, 119, 111, 114, 108, 100])
        }
        );
        console.log(sig);
    } catch(e) {
        alert(e)
    }
});

// Display single-sig BTC address on device for verification
const simpleBtn = document.getElementById("btcAddressSimple");
simpleBtn.addEventListener("click", async () => {
    const address = await device.api.btcDisplayAddressSimple(
        constants.messages.BTCCoin.BTC,
        getKeypathFromString("m/49'/0'/0'/0/0"),
        constants.messages.BTCScriptConfig_SimpleType.P2WPKH_P2SH,
        false,
    );
    simpleBtn.textContent = "Confirm the following address on the BitBox:\n" + address;
    simpleBtn.className = "btn btn-warning";
    try {
        const addressConfirmed = await device.api.btcDisplayAddressSimple(
            constants.messages.BTCCoin.BTC,
            getKeypathFromString("m/49'/0'/0'/0/0"),
            constants.messages.BTCScriptConfig_SimpleType.P2WPKH_P2SH,
        );
        simpleBtn.textContent = "Success!:\n" + addressConfirmed;
        simpleBtn.className = "btn btn-success";
    } catch (e) {
        simpleBtn.textContent = JSON.stringify(e);
        simpleBtn.className = "btn btn-danger";
    }
});

// Mock sample single-sig BTC transaction
function makeExampleTx(keypathAccount) {
    const inputs = [
        {
            "prevOutHash": Uint8Array.from([0xc5, 0x8b, 0x7e, 0x3f, 0x12, 0x00, 0xe0, 0xc0, 0xec, 0x9a, 0x5e, 0x81, 0xe9, 0x25, 0xba, 0xfa, 0xce, 0x2c, 0xc1, 0xd4, 0x71, 0x55, 0x14, 0xf2, 0xd8, 0x20, 0x5b, 0xe2, 0x50, 0x8b, 0x48, 0xee]),
            "prevOutIndex": 1,
            "prevOutValue": "60005000", // satoshis as a decimal string
            "sequence": 0xFFFFFFFF,
            "keypath": keypathAccount.concat([0, 0]),
            "prevTx": {
                "version": 1,
                "locktime": 0,
                "inputs": [
                    {
                        "prevOutHash": new Uint8Array(32).fill(49), // arbitrary constant,
                        "prevOutIndex": 0,
                        "signatureScript": new TextEncoder().encode("some signature script"),
                        "sequence": 0xFFFFFFFF,
                    }
                ],
                "outputs": [{
                    "value": "60005000",
                    "pubkeyScript": new TextEncoder().encode("some pubkey script"),
                }],
            },
        },
        {
            "prevOutHash": Uint8Array.from([0xc5, 0x8b, 0x7e, 0x3f, 0x12, 0x00, 0xe0, 0xc0, 0xec, 0x9a, 0x5e, 0x81, 0xe9, 0x25, 0xba, 0xfa, 0xce, 0x2c, 0xc1, 0xd4, 0x71, 0x55, 0x14, 0xf2, 0xd8, 0x20, 0x5b, 0xe2, 0x50, 0x8b, 0x48, 0xee]),
            "prevOutIndex": 1,
            "prevOutValue": "60005000", // satoshis as a decimal string
            "sequence": 0xFFFFFFFF,
            "keypath": keypathAccount.concat([0, 1]),
            "prevTx": {
                "version": 1,
                "locktime": 0,
                "inputs": [
                    {
                        "prevOutHash": new Uint8Array(32).fill(49), // arbitrary constant,
                        "prevOutIndex": 0,
                        "signatureScript": new TextEncoder().encode("some signature script"),
                        "sequence": 0xFFFFFFFF,
                    }
                ],
                "outputs": [{
                    "value": "60005000",
                    "pubkeyScript": new TextEncoder().encode("some pubkey script"),
                }],
            },
        }
    ];
    const outputs = [
        {
            "ours": true, // change
            "keypath": keypathAccount.concat([1, 0]),
            "value": "100000000", // satoshis as a decimal string
        },
        {
            "ours": false,
            "type": constants.messages.BTCOutputType.P2WSH,
            "hash": new Uint8Array(32).fill(49), // arbitrary constant
            "value": "20000000", // satoshis as a decimal string,
        },
    ];
    const version = 1;
    const locktime = 0;
    return { inputs, outputs, version, locktime };
}

// Sign single-sig BTC transaction
document.getElementById("btcSignSimple").addEventListener("click", async () => {
    const keypathAccount = getKeypathFromString("m/84'/0'/0'");
    const { inputs, outputs, version, locktime } = makeExampleTx(keypathAccount);
    try {
        const signatures = await device.api.btcSignSimple(
            constants.messages.BTCCoin.BTC,
            constants.messages.BTCScriptConfig_SimpleType.P2WPKH,
            keypathAccount,
            inputs,
            outputs,
            version,
            locktime,
        );
        console.log("Signatures: ", signatures);
    } catch(err) {
        console.log(err);
        if (isErrorAbort(err)) {
            alert("aborted by user");
        } else {
            alert(err.Message);
        }
    }
});

// Sign single-sig BTC transaction
document.getElementById("btcSignMsg").addEventListener("click", async () => {
    try {
        const result = await device.api.btcSignMessage(
            constants.messages.BTCCoin.BTC,
            constants.messages.BTCScriptConfig_SimpleType.P2WPKH,
            getKeypathFromString("m/84'/0'/0'/0/0"),
            // "hello world"
            new Uint8Array([104, 101, 108, 108, 111, 32, 119, 111, 114, 108, 100]),
        );
        alert([
            "Signature:",
            toHex(result.signature),
            "Signature (Electrum):" ,
            toBase64(result.electrumSignature),
        ].join("\n"));
        console.log(result);
    } catch(err) {
        console.log(err);
        if (isErrorAbort(err)) {
            alert("aborted by user");
        } else {
            alert(err.Message);
        }
    }
});

// Mock sample multi-sig BTC transaction
async function makeExampleMultisigAccount(keypathAccount) {
    const coin = constants.messages.BTCCoin.BTC;

    const ourXPub = await device.api.btcXPub(coin, keypathAccount, constants.messages.BTCXPubType.ZPUB, false);

    // One of the xpubs must be ours.
    // The xpubs can be provided in any version (xpub, ypub, zpub, etc.).
    const xpubs = [
        ourXPub,
        "Zpub75oAskzNv1fFuPc4ue6dVyZETFdroU4Bw2NNTL6Yz8WzLx3athVwRKmuMoSmwqXFL5jw5ZEdbinsW5TSEiHQTFUuZg4Pk2v2mma1ERpHWgq",
    ];

    const account = {
        "coin": coin,
        "keypathAccount": keypathAccount,
        "threshold": 1,
        "xpubs": xpubs,
        "ourXPubIndex": 0,
    };

    const getName = async () => prompt("Please give a name for the multisig account (max 30 chars)");
    await device.api.btcMaybeRegisterScriptConfig(account, getName);
    return account;
}

// Display single-sig BTC address on device for verification
document.getElementById("btcAddressMultisig").addEventListener("click", async () => {
    const keypath = getKeypathFromString("m/48'/0'/0'/2'/0/0");
    const keypathAccount = keypath.slice(0, keypath.length - 2);

    try {
        const account = await makeExampleMultisigAccount(keypathAccount);
        await device.api.btcDisplayAddressMultisig(account, keypath);
    } catch(err) {
        if (isErrorAbort(err)) {
            alert("aborted by user");
        } else {
            alert(err.Message);
        }
    }
});

// Sign multi-sig BTC transaction
document.getElementById("btcSignMultisig").addEventListener("click", async () => {
    const coin = constants.messages.BTCCoin.BTC;
    const keypathAccount = getKeypathFromString("m/48'/0'/0'/2'");

    const account = await makeExampleMultisigAccount(keypathAccount);

    const { inputs, outputs, version, locktime } = makeExampleTx(keypathAccount);

    try {
        const signatures = await device.api.btcSignMultisig(
            account,
            inputs,
            outputs,
            version,
            locktime,
        );
        console.log("Signatures: ", signatures);
    } catch(err) {
        if (isErrorAbort(err)) {
            alert("aborted by user");
        } else {
            alert(err.Message);
        }
    }
});

const cardanoGetAddress = async (display) => device.api.cardanoAddress(
  constants.messages.CardanoNetwork.CardanoMainnet,
  {
    pkhSkh: {
      keypathPayment: getKeypathFromString("m/1852'/1815'/0'/0/0"),
      keypathStake: getKeypathFromString("m/1852'/1815'/0'/2/0"),
    },
  }, display);

cardanoAddr.addEventListener("click", async () => {
    try {
        const address = await cardanoGetAddress(false);
        cardanoAddr.textContent = "Confirm the following address on the BitBox:\n" + address;
        cardanoAddr.className = "btn btn-warning";

        await cardanoGetAddress(true);
        cardanoAddr.textContent = "Success!:\n" + address;
        cardanoAddr.className = "btn btn-success";
    } catch (e) {
        cardanoAddr.textContent = JSON.stringify(e);
        cardanoAddr.className = "btn btn-danger";
    }
});

cardanoPub.addEventListener("click", async () => {
    const xpubs = await device.api.cardanoXPubs([
        getKeypathFromString("m/1852'/1815'/0'"),
        getKeypathFromString("m/1852'/1815'/1'"),
    ]);
    alert("Account #0\n" + toHex(xpubs[0]));
    alert("Account #1\n" + toHex(xpubs[1]));
});


cardanoSign.addEventListener("click", async () => {
    const ttl = "41115811";
    const fee = "170499";
    const validityIntervalStart = "41110811";
    const response = await device.api.cardanoSignTransaction({
        network: constants.messages.CardanoNetwork.CardanoMainnet,
        inputs: [{
            "keypath": getKeypathFromString("m/1852'/1815'/0'/0/0"),
            "prevOutHash": fromHex("59864ee73ca5d91098a32b3ce9811bac1996dcbaefa6b6247dcaafb5779c2538"),
            "prevOutIndex": 0,
        }],
        outputs: [{
            "encodedAddress": "addr1q9qfllpxg2vu4lq6rnpel4pvpp5xnv3kvvgtxk6k6wp4ff89xrhu8jnu3p33vnctc9eklee5dtykzyag5penc6dcmakqsqqgpt",
            "value": "1000000",
        },
         {
             "encodedAddress": await cardanoGetAddress(false),
             "value": "4829501",
             "scriptConfig": {
                 "pkhSkh": {
                     "keypathPayment": getKeypathFromString("m/1852'/1815'/0'/0/0"),
                     "keypathStake": getKeypathFromString("m/1852'/1815'/0'/2/0"),
                 },
             },
        }],
        fee,
        ttl,
        certificates: [
            { "stakeRegistration": {
                "keypath": getKeypathFromString("m/1852'/1815'/0'/2/0"),
            } },
            { "stakeDelegation": {
                "keypath": getKeypathFromString("m/1852'/1815'/0'/2/0"),
                "poolKeyhash": fromHex("92229dcf782ce8a82050fdeecb9334cc4d906c6eb66cdbdcea86fb5f"),
            } }
        ],
        withdrawals: [{ "keypath": getKeypathFromString("m/1852'/1815'/0'/2/0"), value: "12345" }],
        validityIntervalStart,
    });
    console.log("Sign response:", response);
});
